TypedSyntax/src/node.jl

```diff
@@ -0,0 +1,432 @@
+                        if src.slotnames[i] == Symbol("#unused#")
+                            arg.typ = src.slottypes[i]
+                            i += 1
+                            found = true
+                            break
+                        end
+                        i += 1
+                    end
+                    @assert found
+                    continue
+                elseif nchildren == 2
+                    arg = child(arg, 1)  # extract the name
+                else
+                    error("unexpected number of children: ", children(arg))
+                end
+            end
+            @assert kind(arg) == K"Identifier"
+            argname = arg.val
+            while i <= length(src.slotnames)
+                if src.slotnames[i] == argname
+                    arg.typ = src.slottypes[i]
+                    i += 1
+                    break
+                end
+                i += 1
+            end
+        end
+    end
+    return trootnode
+end
+
+# Recursive construction of the TypedSyntaxNode tree from the SyntaxNodeTree
+function addchildren!(tparent, parent, src::CodeInfo, node2ssa, symtyps, mappings)
+    if haschildren(parent) && tparent.children === nothing
+        tparent.children = TypedSyntaxNode[]
+    end
+    for child in children(parent)
+        tnode = TreeNode(tparent, nothing, TypedSyntaxData(child.data, src, gettyp(node2ssa, child, src)))
+        if tnode.typ === nothing && kind(child) == K"Identifier"
+            tnode.typ = get(symtyps, child, nothing)
+        end
+        push!(tparent, tnode)
+        addchildren!(tnode, child, src, node2ssa, symtyps, mappings)
+    end
+    # In `return f(args..)`, copy any types assigned to `f(args...)` up to the `[return]` node
+    if kind(tparent) == K"return" && haschildren(tparent)
+        tparent.typ = only(children(tparent)).typ
+    end
+    # Replace the entry in `mappings` to be the typed node
+    i = get(node2ssa, parent, nothing)
+    if i !== nothing
+        @assert length(mappings[i]) == 1
+        mappings[i][1] = tparent
+    end
+    return tparent
+end
+
+function gettyp(node2ssa, node, src)
+    i = get(node2ssa, node, nothing)
+    i === nothing && return nothing
+    stmt = src.code[i]
+    if isa(stmt, Core.ReturnNode)
+        arg = stmt.val
+        isa(arg, SSAValue) && return src.ssavaluetypes[arg.id]
+        is_slot(arg) && return src.slottypes[arg.id]
+    end
+    return src.ssavaluetypes[i]
+end
+
+Base.copy(tsd::TypedSyntaxData) = TypedSyntaxData(tsd.source, tsd.typedsource, tsd.raw, tsd.position, tsd.val, tsd.typ)
+
+gettyp(node::AbstractSyntaxNode) = gettyp(node.data)
+gettyp(::JuliaSyntax.SyntaxData) = nothing
+gettyp(data::TypedSyntaxData) = data.typ
+
+# function sparam_name(mi::MethodInstance, i::Int)
+#     sig = (mi.def::Method).sig::UnionAll
+#     while true
+#         i == 1 && break
+#         sig = sig.body::UnionAll
+#         i -= 1
+#     end
+#     return sig.var.name
+# end
```

Comment:
@aviatesk, I didn't find a need to wire this in with the new design, but if you know of one, please provide a (failing) test case!

Reply:
I'd argue for `(myzero(x::Float64::T) where T)::Float64` or `(myzero((x::Float64)::T) where T)::Float64`; I think "binding" the `Float64` more tightly to `x` than the type restriction makes sense; after all `Float64` is the inferred type of `x`, and printing the type in the inner-most operation makes sense. This choice treats `::` basically as any other infix operator.

Reply:
Okay. Now I think `(myzero((x::Float64)::T) where T)::Float64` is the most reasonable option but I would like to leave the decision to you.